File: server/src/collectors/jmaOceanWave.js
/**
 * JMA Ocean Wave Collector
 *
 * Real upstream (scraped from the bosai/wave web app):
 *   - Stations:  https://www.jma.go.jp/bosai/wave/const/pointinfo.json
 *   - Per-stn:   https://www.jma.go.jp/bosai/wave/data/swjp/<number>.json
 *
 * Per-station JSON is a chronological array of hourly readings:
 *   { id, date (UTC), height (cm), period (s) }
 *
 * Falls back to a small seed of major coastal buoys if the JMA endpoints
 * fail (e.g. JMA reorganises the bosai pages again).
 */

const POINTINFO_URL = 'https://www.jma.go.jp/bosai/wave/const/pointinfo.json';
const PER_STATION_URL = (n) => `https://www.jma.go.jp/bosai/wave/data/swjp/${n}.json`;

const SEED_WAVE_BUOYS = [
  { name: '釧路沖', lat: 42.7000, lon: 144.7000, height_m: 2.1, period_s: 7.5, region: 'Pacific North' },
  { name: '銚子沖', lat: 35.7000, lon: 140.9500, height_m: 1.7, period_s: 6.8, region: 'Pacific Central' },
  { name: '潮岬沖', lat: 33.4000, lon: 135.7500, height_m: 1.8, period_s: 7.0, region: 'Pacific Central' },
  { name: '室戸沖', lat: 33.2000, lon: 134.1500, height_m: 1.9, period_s: 7.2, region: 'Pacific Central' },
  { name: '稚内沖', lat: 45.5000, lon: 141.5000, height_m: 1.4, period_s: 6.0, region: 'Japan Sea North' },
  { name: '対馬海峡', lat: 34.2000, lon: 129.5000, height_m: 1.4, period_s: 5.8, region: 'Japan Sea West' },
  { name: '奄美大島', lat: 28.4000, lon: 129.5000, height_m: 1.8, period_s: 7.0, region: 'Nansei' },
  { name: '沖縄本島東', lat: 26.5000, lon: 128.3000, height_m: 1.7, period_s: 7.0, region: 'Okinawa' },
];

async function fetchJsonShort(url, timeoutMs = 8000) {
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), timeoutMs);
    const res = await fetch(url, { signal: ctrl.signal });
    clearTimeout(t);
    if (!res.ok) return null;
    return await res.json();
  } catch { return null; }
}

async function tryJmaWave() {
  const stations = await fetchJsonShort(POINTINFO_URL);
  if (!Array.isArray(stations) || !stations.length) return null;

  const results = await Promise.all(stations.map(async (s) => {
    const series = await fetchJsonShort(PER_STATION_URL(s.number));
    if (!Array.isArray(series) || !series.length) return null;
    // Last entry with a finite height — JMA pads recent rows with nulls
    // for missing observations; walk backwards to find the latest valid one.
    let latest = null;
    for (let i = series.length - 1; i >= 0; i--) {
      const r = series[i];
      if (r && Number.isFinite(r.height)) { latest = r; break; }
    }
    if (!latest) return null;
    const [lat, lon] = Array.isArray(s.latlon) ? s.latlon : [null, null];
    const geocoded = Number.isFinite(lat) && Number.isFinite(lon);
    return {
      type: 'Feature',
      geometry: geocoded ? { type: 'Point', coordinates: [lon, lat] } : null,
      properties: {
        buoy_id: `JMA_WAVE_${s.number}`,
        name: s.name,
        wave_height_m: Math.round(latest.height) / 100, // cm → m
        period_s: latest.period ?? null,
        observed_at: latest.date,
        country: 'JP',
        source: 'jma_bosai_wave',
      },
    };
  }));
  const live = results.filter(Boolean);
  return live.length > 0 ? live : null;
}

function generateSeedData() {
  const now = new Date();
  return SEED_WAVE_BUOYS.map((b, i) => ({
    type: 'Feature',
    geometry: { type: 'Point', coordinates: [b.lon, b.lat] },
    properties: {
      buoy_id: `WAV_${String(i + 1).padStart(5, '0')}`,
      name: b.name,
      wave_height_m: b.height_m,
      period_s: b.period_s,
      region: b.region,
      country: 'JP',
      observed_at: now.toISOString(),
      source: 'jma_wave_seed',
    },
  }));
}

export default async function collectJmaOceanWave() {
  let features = await tryJmaWave();
  const live = !!(features && features.length > 0);
  if (!live) features = generateSeedData();
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: 'jma_ocean_wave',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live,
      live_source: live ? 'jma_bosai_wave' : 'jma_wave_seed',
      description: 'JMA significant wave height observations (bosai/wave) — hourly per-station scraped',
    },
  };
}

File: native/collectors/sources/jma_ocean_wave.c
/* collectors/environment/sources/jma_ocean_wave.c
 * Port of server/src/collectors/jmaOceanWave.js — live JMA bosai/wave path:
 * pointinfo.json (stations) → per-station swjp/<number>.json hourly series →
 * latest finite-height reading → one Feature/station. Seed buoy fallback
 * (rule 7) dropped. */
#include "../../source.h"
#include "../../lib/feedlib.h"
#include "../../lib/geojson.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define POINTINFO "https://www.jma.go.jp/bosai/wave/const/pointinfo.json"

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *stations = feed_get_json(ctx->http, POINTINFO, 8000);
  if (!cJSON_IsArray(stations) || cJSON_GetArraySize(stations) == 0) {
    if (stations) cJSON_Delete(stations);
    return -1;                                      /* tryJmaWave → null */
  }

  cJSON *features = cJSON_CreateArray();
  cJSON *s;
  cJSON_ArrayForEach(s, stations) {
    cJSON *numv = cJSON_GetObjectItem(s, "number");
    char numbuf[32] = {0};
    if (numv && cJSON_IsNumber(numv))
      snprintf(numbuf, sizeof numbuf, "%g", numv->valuedouble);
    else if (numv && cJSON_IsString(numv) && numv->valuestring)
      snprintf(numbuf, sizeof numbuf, "%s", numv->valuestring);
    if (!numbuf[0]) continue;

    char url[160];
    snprintf(url, sizeof url,
      "https://www.jma.go.jp/bosai/wave/data/swjp/%s.json", numbuf);
    cJSON *series = feed_get_json(ctx->http, url, 8000);
    if (!cJSON_IsArray(series) || cJSON_GetArraySize(series) == 0) {
      if (series) cJSON_Delete(series);
      continue;
    }

    /* walk backwards for the last entry with a finite numeric height */
    cJSON *latest = NULL;
    int cnt = cJSON_GetArraySize(series);
    for (int i = cnt - 1; i >= 0; i--) {
      cJSON *r = cJSON_GetArrayItem(series, i);
      cJSON *h = r ? cJSON_GetObjectItem(r, "height") : NULL;
      if (h && cJSON_IsNumber(h) && isfinite(h->valuedouble)) { latest = r; break; }
    }
    if (!latest) { cJSON_Delete(series); continue; }

    cJSON *hv = cJSON_GetObjectItem(latest, "height");
    cJSON *pv = cJSON_GetObjectItem(latest, "period");
    cJSON *dv = cJSON_GetObjectItem(latest, "date");
    cJSON *nm = cJSON_GetObjectItem(s, "name");
    cJSON *ll = cJSON_GetObjectItem(s, "latlon");

    int geocoded = 0;
    double lat = 0, lon = 0;
    if (cJSON_IsArray(ll) && cJSON_GetArraySize(ll) >= 2) {
      cJSON *la = cJSON_GetArrayItem(ll, 0);
      cJSON *lo = cJSON_GetArrayItem(ll, 1);
      if (la && cJSON_IsNumber(la) && lo && cJSON_IsNumber(lo) &&
          isfinite(la->valuedouble) && isfinite(lo->valuedouble)) {
        lat = la->valuedouble; lon = lo->valuedouble; geocoded = 1;
      }
    }

    cJSON *f = cJSON_CreateObject();
    cJSON_AddStringToObject(f, "type", "Feature");
    if (geocoded) {
      cJSON *g = cJSON_CreateObject();
      cJSON_AddStringToObject(g, "type", "Point");
      cJSON *co = cJSON_CreateArray();
      cJSON_AddItemToArray(co, cJSON_CreateNumber(lon));
      cJSON_AddItemToArray(co, cJSON_CreateNumber(lat));
      cJSON_AddItemToObject(g, "coordinates", co);
      cJSON_AddItemToObject(f, "geometry", g);
    } else {
      cJSON_AddItemToObject(f, "geometry", cJSON_CreateNull());
    }

    cJSON *p = cJSON_CreateObject();               /* EXACT JS key order */
    char bid[64];
    snprintf(bid, sizeof bid, "JMA_WAVE_%s", numbuf);
    cJSON_AddStringToObject(p, "buoy_id", bid);
    cJSON_AddItemToObject(p, "name",
      (nm && cJSON_IsString(nm) && nm->valuestring)
        ? cJSON_CreateString(nm->valuestring) : cJSON_CreateNull());
    /* wave_height_m = Math.round(height)/100 */
    double wh = round(hv->valuedouble) / 100.0;
    cJSON_AddNumberToObject(p, "wave_height_m", wh);
    cJSON_AddItemToObject(p, "period_s",
      (pv && cJSON_IsNumber(pv)) ? cJSON_Duplicate(pv, 1) : cJSON_CreateNull());
    cJSON_AddItemToObject(p, "observed_at",
      (dv && cJSON_IsString(dv) && dv->valuestring)
        ? cJSON_CreateString(dv->valuestring) : cJSON_CreateNull());
    cJSON_AddStringToObject(p, "country", "JP");
    cJSON_AddStringToObject(p, "source", "jma_bosai_wave");
    cJSON_AddItemToObject(f, "properties", p);
    cJSON_AddItemToArray(features, f);

    cJSON_Delete(series);
  }
  cJSON_Delete(stations);

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[jma-ocean-wave] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def jma_ocean_wave_def = {
  .id = "jma-ocean-wave", .collector = "environment",
  .name = "JMA Ocean Wave",
  .name_ja = "\xE6\xB0\x97\xE8\xB1\xA1\xE5\xBA\x81 \xE6\xB3\xA2\xE6\xB5\xAA\xE8\xA6\xB3\xE6\xB8\xAC",
   .update_interval_sec = 3600, .run = run };
REGISTER_SOURCE(jma_ocean_wave_def)
